File: py/day24.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

import re

with open("../resources/day24.txt") as f:
    alu_steps, block = [], []
    for i, line in enumerate(f.readlines()):
        if re.match("inp w", line):
            if block:
                alu_steps.append(block)
            block = []
        if i % 18 in [4, 5, 15]:
            val = re.search(r"-?\d+", line)
            block.append(int(val.group()))
    alu_steps.append(block)

def alu_func(z, w, a, b, c):
    x = 0 if w == z % 26 + b else 1
    return z // a * (25 * x + 1) + x * (w + c)

def inverse_alu_func(zprime, a, b, c, mode="max"):
    result = {}
    for w in range(1, 10):
        match = [extra + (a * zprime) for extra in range(0, a)]
        not_match = [extra + (a * (zprime - (w + c))) // 26 for extra in range(0, a)]
        candidates = set([z for z in match + not_match if z > 0])
        for z in candidates:
            if zprime == alu_func(z, w, a, b, c):
                if z in result:
                    result[z] = max(result[z], w) if mode == "max" else min(result[z], w)
                else:
                    result[z] = w
    return result

def table(mode="max"):
    result = {}
    zprimes = [0]
    for digit in range(13, 0, -1):
        group = {}
        for zp in zprimes:
            for z, w in inverse_alu_func(zp, *alu_steps[digit], mode=mode).items():
                if z in group:
                    group[z] = max(group[z], w) if mode == "max" else min(group[z], w)
                else:
                    group[z] = w
        result[digit] = group
        zprimes = group.keys()
    return result

def solve(mode="max"):
    the_table = table(mode=mode)
    candidates = []
    for seed in range(1, 10):
        result = [seed]
        z = 0
        for i, step in enumerate(alu_steps):
            z = alu_func(z, result[-1], *step)
            try:
                result.append(the_table[i+1][z])
            except KeyError:
                break
        if len(result) == 14:
            candidates.append(int("".join([str(x) for x in result])))
    return max(candidates) if mode == "max" else min(candidates)

print("Day 24\n"
      "[Part 1] %d\n"
      "[Part 2] %d"
      %(solve(), solve(mode="min")))

File: src/aoc2021/day24.clj
(ns aoc2021.day24
  (:require [clojure.java.io :as io]))

(comment
 "Day 24: Arithmetic Logic Unit")

(def alu-steps
  "Extracts the three important values from each ALU instruction block."
  (with-open [rdr (io/reader (io/resource "../resources/day24.txt"))]
    (->> (line-seq rdr)
         (vec)
         (partition 18)       ;; Each block is 18 lines long
         (apply map vector)   ;; Transpose -> Each row is now the nth line of all blocks
         (remove #(= (count (set %)) 1)) ;; Remove the lines that are the same for each block
         (apply map vector)   ;; Transpose -> Each row is now the important lines of each block
         (map (fn [block]     ;; Convert the numbers to integers.
                (map #(read-string (re-find #"-?\d+" %))
                     block))))))

(comment
 "The operations do not need to be extracted from the puzzle input because
  the math has been done by hand.")

(defn alu-func
  "Given a z-value, an ALU program block represented by its three unique values
  a, b, and c, and an input value w, returns the resulting value z'."
  [z [a b c] w]
  (let [x  (if (= w (+ (mod z 26) b)) 0 1)  ; x is 0 if w = z mod 26 + b else 1
        y1 (inc (* 25 x))                   ; Z // A * (25x + 1) + x(w + c)
        y2 (* x (+ w c))]
    (+ y2 (quot (* z y1) a))))

(defn inverse-alu-func
  "Given a resulting value z' and an ALU program block, returns a map
  associating each seed-state z to the max/min (`mode`) w value that produces
  that result."
  [z' [a b c] mode]
  (apply merge-with mode
    (for [w (range 1 10)
          ;; match/not-match refers to the condition w = z mod 26 + b
          ;; the additional element % is to reverse the quotient function
          :let [match     (map #(+ % (* a z')) (range a))
                not-match (map #(+ % (quot (* a (- z' (+ w c)))
                                           26))
                               (range a))]
          z (set (filter pos? (concat match not-match)))
          ;; Prune the results by checking them forwards
          :when (= z' (alu-func z [a b c] w))]
      {z w})))

(defn table
  "Creates a table with the information necessary to find the max/min (`mode`)
  MONAD number possible step-by-step. If you are at digit D and the current
  z value is Z, (get-in table [D Z]) gives you the next w value you should input."
  [mode]
  (loop [digit 13, z's [0], result {}]
    (if (neg? digit) result
      (let [solutions (apply merge-with mode
                        (for [z' z's]
                          (inverse-alu-func z' (nth alu-steps digit) mode)))]
        (recur (dec digit) (keys solutions) (conj result {digit solutions}))))))

(defn solve
  "Finds the max/min (`mode`) possible 14-digit MONAD number."
  [mode]
  (let [the-table (table mode)]
    (apply mode
      (remove nil?
        (for [seed (range 1 10)]
          (loop [[step & steps] alu-steps
                 result         [seed]
                 z              0]
            (let [z' (alu-func z step (peek result))]
              (if steps
                (when-let [w (get-in the-table [(count result) z'])]
                  (recur steps (conj result w) z'))
                (read-string (apply str result))))))))))

(time
 (do
   (println "Day 24: Arithmetic Logic Unit")
   (println "[Part 1]" (solve max))
   (println "[Part 2]" (solve min))))
